// patternfly-react/dist/esm/components/TreeView/TreeViewNode.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { noop } from '../../common/helpers';
import TreeViewExpand from './TreeViewExpand';
import TreeViewIcon from './TreeViewIcon';
import TreeViewIndents from './TreeViewIndents';

var TreeViewNode = function (_Component) {
  _inherits(TreeViewNode, _Component);

  function TreeViewNode(props) {
    _classCallCheck(this, TreeViewNode);

    // A node can be set to be expanded by default
    var _this = _possibleConstructorReturn(this, (TreeViewNode.__proto__ || Object.getPrototypeOf(TreeViewNode)).call(this, props));

    _this.state = {
      expanded: props.node.hasOwnProperty('state') && props.node.state.expanded || false
    };

    _this.toggleExpand = _this.toggleExpand.bind(_this);
    _this.handleSelect = _this.handleSelect.bind(_this);
    return _this;
  }

  // Collapse the current node if any of its parents is collapsed. This should
  // only fire for nodes that are level 2 or greater


  _createClass(TreeViewNode, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if (!nextProps.visible && nextProps.level > 1) {
        this.setState(function () {
          return { expanded: false };
        });
      }
    }
  }, {
    key: 'handleSelect',
    value: function handleSelect() {
      var _props = this.props,
          node = _props.node,
          selectNode = _props.selectNode;

      if (node.selectable) {
        selectNode(node);
      }
    }
  }, {
    key: 'toggleExpand',
    value: function toggleExpand(e) {
      e.stopPropagation();
      this.setState(function (prevState) {
        return { expanded: !prevState.expanded };
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          node = _props2.node,
          level = _props2.level,
          visible = _props2.visible,
          selectNode = _props2.selectNode;
      var expanded = this.state.expanded;

      var classes = classNames('list-group-item', {
        'node-hidden': level > 1 ? !visible : false,
        'node-selected': node.selected
      });
      return React.createElement(
        React.Fragment,
        null,
        React.createElement(
          'li',
          { className: classes, onClick: this.handleSelect },
          React.createElement(TreeViewIndents, { level: level }),
          React.createElement(TreeViewExpand, {
            nodes: node.nodes,
            expanded: expanded,
            toggleExpand: this.toggleExpand
          }),
          React.createElement(TreeViewIcon, { icon: node.icon }),
          node.text
        ),
        node.nodes && node.nodes.map(function (childNode, index) {
          return React.createElement(TreeViewNode, {
            node: childNode,
            key: index,
            level: level + 1,
            visible: expanded,
            selectNode: selectNode
          });
        })
      );
    }
  }]);

  return TreeViewNode;
}(Component);

TreeViewNode.propTypes = {
  node: PropTypes.object,
  level: PropTypes.number.isRequired,
  visible: PropTypes.bool,
  selectNode: PropTypes.func
};

TreeViewNode.defaultProps = {
  node: {},
  visible: false,
  selectNode: noop
};

export default TreeViewNode;
