// patternfly-react/dist/esm/components/Masthead/MastheadDropdown.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { DropdownButton } from '../Button';

/**
 * Masthead
 */
var MastheadDropdown = function MastheadDropdown(_ref) {
  var className = _ref.className,
      id = _ref.id,
      title = _ref.title,
      children = _ref.children,
      props = _objectWithoutProperties(_ref, ['className', 'id', 'title', 'children']);

  var dropdownClasses = classNames('dropdown-toggle', 'nav-item-iconic', className);

  return React.createElement(
    'li',
    _extends({ className: 'dropdown' }, props),
    React.createElement(
      DropdownButton,
      {
        id: id,
        className: dropdownClasses,
        bsStyle: 'link',
        title: title
      },
      children
    )
  );
};

MastheadDropdown.propTypes = {
  /** Additional element css classes */
  className: PropTypes.string,
  /** ID of dropdown */
  id: PropTypes.string.isRequired,
  /** Title of dropdown */
  title: PropTypes.node,
  /** Children (likely MenuItem's) */
  children: PropTypes.node
};

MastheadDropdown.defaultProps = {
  className: '',
  title: '',
  children: null
};

export default MastheadDropdown;
