// patternfly-react/dist/js/components/Masthead/MastheadDropdown.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Button = require('../Button');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/**
 * Masthead
 */
var MastheadDropdown = function MastheadDropdown(_ref) {
  var className = _ref.className,
      id = _ref.id,
      title = _ref.title,
      children = _ref.children,
      props = _objectWithoutProperties(_ref, ['className', 'id', 'title', 'children']);

  var dropdownClasses = (0, _classnames2.default)('dropdown-toggle', 'nav-item-iconic', className);

  return _react2.default.createElement(
    'li',
    _extends({ className: 'dropdown' }, props),
    _react2.default.createElement(
      _Button.DropdownButton,
      {
        id: id,
        className: dropdownClasses,
        bsStyle: 'link',
        title: title
      },
      children
    )
  );
};

MastheadDropdown.propTypes = {
  /** Additional element css classes */
  className: _propTypes2.default.string,
  /** ID of dropdown */
  id: _propTypes2.default.string.isRequired,
  /** Title of dropdown */
  title: _propTypes2.default.node,
  /** Children (likely MenuItem's) */
  children: _propTypes2.default.node
};

MastheadDropdown.defaultProps = {
  className: '',
  title: '',
  children: null
};

exports.default = MastheadDropdown;
