// patternfly-react/dist/esm/components/FieldLevelHelp/FieldLevelHelp.js
function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

import React from 'react';
import PropTypes from 'prop-types';
import { Icon } from '../Icon';
import { Button } from '../Button';
import { Popover } from '../Popover';
import { OverlayTrigger } from '../OverlayTrigger';

/**
 * FieldLevelHelp Component for Patternfly React
 */
var FieldLevelHelp = function FieldLevelHelp(_ref) {
  var children = _ref.children,
      content = _ref.content,
      close = _ref.close,
      props = _objectWithoutProperties(_ref, ['children', 'content', 'close']);

  var trigger = 'click';
  var overlay = React.createElement(
    Popover,
    { id: 'popover' },
    content
  );
  var rootClose = close === 'true';

  return React.createElement(
    OverlayTrigger,
    {
      overlay: overlay,
      placement: 'top',
      trigger: trigger.split(' '),
      rootClose: rootClose
    },
    React.createElement(
      Button,
      { bsStyle: 'link', className: 'popover-pf-info' },
      React.createElement(Icon, { type: 'pf', name: 'info' })
    )
  );
};

FieldLevelHelp.propTypes = {
  /** additional fieldlevelhelp classes */
  content: PropTypes.node,
  /** leave popover/tooltip open  */
  close: PropTypes.string,
  /** children nodes  */
  children: PropTypes.node
};
FieldLevelHelp.defaultProps = {
  content: null,
  close: 'true',
  children: null
};

export default FieldLevelHelp;
