// patternfly-react/dist/esm/components/Wizard/Wizard.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

import React from 'react';
import PropTypes from 'prop-types';
import { noop, Modal } from '../../index';

/**
 * Wizard - main Wizard component.
 */
var Wizard = function Wizard(_ref) {
  var children = _ref.children,
      className = _ref.className,
      dialogClassName = _ref.dialogClassName,
      show = _ref.show,
      onClose = _ref.onClose,
      onExited = _ref.onExited,
      rest = _objectWithoutProperties(_ref, ['children', 'className', 'dialogClassName', 'show', 'onClose', 'onExited']);

  return React.createElement(
    Modal,
    _extends({
      show: show,
      onHide: onClose,
      onExited: onExited,
      dialogClassName: dialogClassName || 'modal-lg wizard-pf'
    }, rest),
    React.createElement(
      'div',
      { className: className },
      children
    )
  );
};
Wizard.propTypes = {
  /** Children nodes */
  children: PropTypes.node,
  /** Additional css classes */
  className: PropTypes.string,
  /** Wizard dialog class */
  dialogClassName: PropTypes.string,
  /** show modal */
  show: PropTypes.bool,
  /** on close callback */
  onClose: PropTypes.func,
  /** on exited callback */
  onExited: PropTypes.func
};
Wizard.defaultProps = {
  children: null,
  className: '',
  dialogClassName: '',
  show: false,
  onClose: noop,
  onExited: noop
};
export default Wizard;
